### ISSUE TITLE
strictTemplates mode emits confusing type checking error for two-way data-binding with missing @Output declaration (Type 'Event' is not assignable to type X)

### ISSUE THREAD

**Issue by IgorMinar** [angular/angular#39300]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/fw-compiler 

### Is this a regression?

 No

### Description

With `strictTemplates` mode turned on, regulard two-way data-binding no longer passes type-check.

It appears that we don't special retain the information about the type of the value assigned via `(xyzChange)="prop = assignedValue` and instead consider it to be the generic `Event` type.

## 🔬 Minimal Reproduction

https://github.com/aikidave/two-way-binding/blob/c2dc23fdc6120c13b9776e4be44516ab1f2e34f4/src/app/parent/parent.component.html#L5

## 🔥 Exception or Error

```
<app-child [(value)]="counterValue"></app-child>
```
fails with:
```
  ERROR in src/app/parent/parent.component.html:5:23 - error TS2322: Type 'Event' is not assignable to type 'number'.

5 <app-child [(value)]="counterValue"></app-child>
```

## 🌍  Your Environment

**Angular Version:**
<pre><code>

Angular CLI: 10.0.8
Node: 12.13.1
OS: darwin x64

Angular: 10.0.14
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.1000.8
@angular-devkit/build-angular     0.1000.8
@angular-devkit/build-optimizer   0.1000.8
@angular-devkit/build-webpack     0.1000.8
@angular-devkit/core              10.0.8
@angular-devkit/schematics        10.0.8
@angular/cli                      10.0.8
@ngtools/webpack                  10.0.8
@schematics/angular               10.0.8
@schematics/update                0.1000.8
rxjs                              6.5.5
typescript                        3.9.7
webpack                           4.43.0
</code></pre>

**Anything else relevant?**

I tried various workarounds and got mixed results:

```
<app-child [(value)]="$any(counterValue)"></app-child>
```
fails with:

```
  ERROR in Template parse errors:
Parser Error: Unexpected token '=' at column 19 in [$any(counterValue)=$event] in /Users/iminar/tmp/two-way-binding/src/app/parent/parent.component.html@10:22 ("
```

decomposing the binding as:
```
app-child [value]="counterValue" (valueChange)="counterValue=$event"></app-child>
```
fails with:
```
  ERROR in src/app/parent/parent.component.html:16:50 - error TS2322: Type 'Event' is not assignable to type 'number'.

16 <app-child [value]="counterValue" (valueChange)="counterValue=$event"></app-child>
```

and finally, this is the only version I was able to get to work with `strictTemplates`:
```
<app-child [value]="counterValue" (valueChange)="counterValue=$any($event)"></app-child>
```

Lastly, the two-way databinding and `strictTemplates` seem to have another issue that might be somewhat related:  https://github.com/angular/angular/issues/36551

**Comment by IgorMinar** [comment]
@JoostK pointed out that the application has a bug where the child component doesn't properly define `@Output`.

The following patch fixes the issue:

```
diff --git a/src/app/child/child.component.ts b/src/app/child/child.component.ts
index d354898..6e649c9 100644
--- a/src/app/child/child.component.ts
+++ b/src/app/child/child.component.ts
@@ -8,7 +8,7 @@ import { Component, Input, Output, EventEmitter, OnInit } from '@angular/core';
 export class ChildComponent implements OnInit {
 
   @Input() value = 0;
-  @Output() incrementValue = new EventEmitter<number>();
+  @Output() valueChange = new EventEmitter<number>();
 
   constructor() { }
 
@@ -17,7 +17,7 @@ export class ChildComponent implements OnInit {
 
   increase(): void {
     this.value++;
-    this.incrementValue.emit(this.value);
+    this.valueChange.emit(this.value);
   }
 
 }
```

What however remains is that the original error is confusing and misleading.

In the absence of the proper @Output, the compiler assumes that the `(valueChange)` our the outward part of the `[(value)]` binding must be a binding to a DOM event. Hence the `Event` type we are trying to typecheck against. This is however not the case at all and it hides the real issue — the missing output declaration in the child component.

Unless the element we are binding to is a web component, or a native element then we shouldn't fall back to assuming that the unknown event is a DOM event. We should simply error and say that binding to the `value` output is incorrect because such output does not exist on the component.

**Comment by danieltre23** [comment]
Fixed in #42601

**Comment by PieterjanDeClippel** [comment]
Thank you. I did forget the `@Output` decorator...
